models/rate_neuron_ipn.h

```diff
@@ -0,0 +1,326 @@
+#include "ring_buffer.h"
+#include "recordables_map.h"
+#include "universal_data_logger.h"
+
+#include <string>
+
+namespace nest
+{
+/**
+ * Base class for rate model with input noise.
+ *
+ * This class needs to be instantiated with a gain function.
+ *
+ * References:
+ *
+ * Hahne, J., Dahmen, D., Schuecker, J., Frommer, A.,
+ * Bolten, M., Helias, M. and Diesmann, M. (2017).
+ * Integration of Continuous-Time Dynamics in a
+ * Spiking Neural Network Simulator.
+ * Front. Neuroinform. 11:34. doi: 10.3389/fninf.2017.00034
+ *
+ * @see lin_rate, tanh_rate, thresholdlin_rate
+ */
+
+template < class TGainfunction >
+class rate_neuron_ipn : public Archiving_Node
+{
+
+public:
+  typedef Node base;
+
+  rate_neuron_ipn();
+  rate_neuron_ipn( const rate_neuron_ipn& );
+
+  /**
+   * Import sets of overloaded virtual functions.
+   * @see Technical Issues / Virtual Functions: Overriding, Overloading, and
+   * Hiding
+   */
+  using Node::handle;
+  using Node::sends_secondary_event;
+
+  void handle( RateNeuronEvent& );
+  void handle( DelayRateNeuronEvent& );
+  void handle( DataLoggingRequest& );
+
+  port handles_test_event( RateNeuronEvent&, rport );
+  port handles_test_event( DelayRateNeuronEvent&, rport );
+  port handles_test_event( DataLoggingRequest&, rport );
+
+  void
+  sends_secondary_event( RateNeuronEvent& )
+  {
+  }
+  void
+  sends_secondary_event( DelayRateNeuronEvent& )
+  {
+  }
+
+
+  void get_status( DictionaryDatum& ) const;
+  void set_status( const DictionaryDatum& );
+
+private:
+  void init_state_( const Node& proto );
+  void init_buffers_();
+  void calibrate();
+
+  TGainfunction gain_;
+
+  bool update_( Time const&, const long, const long, const bool );
+
+  void update( Time const&, const long, const long );
+  bool wfr_update( Time const&, const long, const long );
+
+  // The next two classes need to be friends to access the State_ class/member
+  friend class RecordablesMap< rate_neuron_ipn< TGainfunction > >;
+  friend class UniversalDataLogger< rate_neuron_ipn< TGainfunction > >;
+
+  // ----------------------------------------------------------------
+
+  /**
+   * Independent parameters of the model.
+   */
+  struct Parameters_
+  {
+
+    /** Time constant in ms. */
+    double tau_;
+
+    /** Gaussian white noise standard deviation. */
+    double std_;
+
+    /** Gaussian white noise mean.*/
+    double mean_;
+
+    /** Target of non-linearity.
+        True: Gain function applied to linearly summed input.
+        False: Gain function applied to each input before summation.
+    **/
+    bool linear_summation_;
+
+    Parameters_(); //!< Sets default parameter values
+
+    void get( DictionaryDatum& ) const; //!< Store current values in dictionary
+
+    void set( const DictionaryDatum& );
+  };
+
+  // ----------------------------------------------------------------
+
+  /**
+   * State variables of the model.
+   */
+  struct State_
+  {
+    double r_; //!< Rate
+    double x_; //!< Noise
```

Comment:
Good suggestion. We changed the names 
r_ -> rate_
x_ -> noise_
y_ -> noisy_rate_